### PR TITLE
Clan sync

### LINKED ISSUE #851: Sync: Clan Fields
## Description

1. Sync the values in the **Synced** list below.
2. Create an end to end (E2E) tests for all fields (1 for server changing the value and 1 for client changing the value if applicable)
3. Create a command to test changing each value in game [see similar command](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/blob/development/source/GameInterface/Services/Towns/Commands/TownDebugCommand.cs#L455)

Also see additional information at the bottom for more information and templates

**Key**
Server side only (only allow the running of the function on the server side)
Client side only (only allow the running of the function on the client side)
Client side sync (request the server to change the value, server then changes the value and replies to the client allowing the change)
Server side sync (only server allows running of the function and send to all clients the value changed and clients update the value/call original function)

**Synced**

| Method/Field Name | Sync Type | Additional Information |
|:-----------------|:---------:|:--------------|
| _isEliminated | Server side sync | |
| _kingdom | Server side sync | |
| _influence | Server side sync | |
| _clanMidSettlement |  Server side sync | Might be a lot of messages and not needed |
| _basicTroop | Server side sync | |
| _leader | Server side sync | |
| _banner | Server side sync | |
| _tier | Server side sync | |
| _aggressiveness | Server side sync | |
| _tributeWallet | Server side sync | |
| _home | Server side sync | |
| _clanDebtToKingdom | Server side sync | |

**Deferred**

None

**Externally Deferred**

None

**Non-Synced**

| Method/Field Name | Remarks |
|:-----------------|:---------:|
| _defaultPartyTemplate | Only set on load |
| _distanceToClosestNonAllyFortificationCache | Not used |
| _distanceToClosestNonAllyFortificationCacheDirty | Not used |

## Intended Design

For the patch itself please refer to https://github.com/Bannerlord-Coop-Team/BannerlordCoop/wiki/Patching-Basics#field-patching

**Field Sync Dataflow**
FieldSyncDataflow
**Integration Testing**
image
## Location

Create a branch based from [development](https://github.com/Bannerlord-Coop-Team/BannerlordCoop)

## Related Issues

Blocked by:
- #850

## Requirements

N/A

## Additional information

Using the [GameInterface Service Tempale](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/tree/development/source/GameInterface/Services/Template) create commands to test the synced values.

## Definition of Done
- [ ] Class level comments exist for all new classes.
- [ ] XUnit E2E tests exist for every property on the server (and client if applicable)
- [ ] Each field has a command getter and setter that used the correct intercept

### LINKED ISSUE #852: Sync: Clan Properties
## Description

1. Sync the values in the **Synced** list below.
2. Create an end to end (E2E) tests for all properties (1 for server changing the value and 1 for client changing the value if applicable)
3. Create a command to test changing each property in game [see similar command](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/blob/development/source/GameInterface/Services/Towns/Commands/TownDebugCommand.cs#L455)

Also see additional information at the bottom for more information and templates

**Key**
Server side only (only allow the running of the function on the server side)
Client side only (only allow the running of the function on the client side)
Client side sync (request the server to change the value, server then changes the value and replies to the client allowing the change)
Server side sync (only server allows running of the function and send to all clients the value changed and clients update the value/call original function)

**Synced**

| Method/Field Name | Sync Type | Additional Information |
|:-----------------|:---------:|:--------------|
| Name | Server side sync | |
| InformalName | Server side sync | |
| Culture | Server side sync | |
| LastFactionChangeTime | Server side sync | |
| AutoRecruitmentExpenses | Server side sync | |
| IsNoble | Server side sync | |
| TotalStrength | Server side sync | This might have a lot of messages |
| MercenaryAwardMultiplier | Server side sync | |
| LabelColor | Server side sync | |
| InitialPosition | Server side sync |
| IsRebelClan | Server side sync | |
| IsUnderMercenaryService | Server side sync | |
| Color | Server side sync | |
| Color2 | Server side sync | |
| BannerBackgroundColorPrimary | Server side sync | |
| BannerBackgroundColorSecondary | Server side sync | |
| BannerIconColor | Server side sync | |
| _midPointCalculated | Server side sync | This might have a lot of messages |
| Renown | Server side sync | |
| NotAttackableByPlayerUntilTime | Server side sync | Might be used to prevent ai from attacking other players |

**Deferred**

| Method/Field Name | Remarks |
|:-----------------|:---------:|
| Kingdom | Synced by _kingdom |
| Influence | Synced by _influence |
| BasicTroop | Synced by _basicTroop |
| Banner | Synced by _banner |
| Tier | Synced by _tier |
| Aggressiveness | Synced by _aggressiveness |
| TributeWallet | Synced by _tributeWallet |
| HomeSettlement | Synced by _home |
| DebtToKingdom | Synced by _clanDebtToKingdom |

**Externally Deferred**

None

**Non-Synced**

| Method/Field Name | Remarks |
|:-----------------|:---------:|
| EncyclopediaText | Only set on load |
| IsMinorFaction | Only set on load |
| IsOutlaw | Only set on load |
| IsNomad | Only set on load |
| IsMafia | Only set on load |
| IsClanTypeMercenary | Only set on load |
| IsSect | Only set on load |
| AlternativeColor | Only set on load |
| AlternativeColor2 | Only set on load |
| IsBanditFaction | Only set on load |
| MainHeroCrimeRating | Not used |

## Intended Design

You are expected to only sync the property setters.
Try using the new [autosync tool](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/wiki/Property-AutoSync)

## Location

Create a branch based from [development](https://github.com/Bannerlord-Coop-Team/BannerlordCoop)

## Related Issues

Blocked by:
- #850

## Requirements

N/A

## Additional information

Using the [GameInterface Service Tempale](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/tree/development/source/GameInterface/Services/Template) create commands to test the synced values.

## Definition of Done
- [ ] Class level comments exist for all new classes.
- [ ] XUnit E2E tests exist for every property on the server (and client if applicable)
- [ ] Each property has a command getter and setter

### PR DESCRIPTION
**Pull request by EgardA**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] All new classes have class-level documentation comments, if there are any at all
- [ ] Tests for the changes have been added (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation update
- [ ] Other... Please describe:

## What is the new behavior?

Resolves #851
Resolves #852

### CODE REVIEW

File: source/E2E.Tests/Services/Clans/ClanSyncTests.cs [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+        private string HeroId;
+
+        public ClanSyncTests(ITestOutputHelper output)
+        {
+            TestEnvironment = new E2ETestEnvironment(output);
+
+            var kingdom = GameObjectCreator.CreateInitializedObject<Kingdom>();
+            var settlement = GameObjectCreator.CreateInitializedObject<Settlement>();
+            var characterObject = GameObjectCreator.CreateInitializedObject<CharacterObject>();
+            var hero = GameObjectCreator.CreateInitializedObject<Hero>();
```
Comment (garrettluskey) [inline source/E2E.Tests/Services/Clans/ClanSyncTests.cs:42]: These should be separate instances on each client and the server

**Comment by garrettluskey** [comment]
Kindly add which issue is resolve

File: source/E2E.Tests/Services/Clans/ClanSyncTests.cs [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+                homeIntercept.Invoke(null, new object[] { serverClan, serverSettlement });
+                clanDebtToKingdomIntercept.Invoke(null, new object[] { serverClan, 25 });
+            });
+
+            foreach (var client in TestEnvironment.Clients)
+            {
+                Assert.True(client.ObjectManager.TryGetObject(clanId, out Clan clientClan));
+                Assert.Equal(serverClan.Name.Value, clientClan.Name.Value);
+                Assert.Equal(serverClan.InformalName.Value, clientClan.InformalName.Value);
+                Assert.Equal(serverClan.Culture, clientClan.Culture);
```
Comment (garrettluskey) [inline source/E2E.Tests/Services/Clans/ClanSyncTests.cs]:
```suggestion
                Assert.Equal(serverClan.Culture.StringId, clientClan.Culture.StringId);
```

File: source/GameInterface/Services/Clans/Handlers/ClanAdoptionHandler.cs [hunk 3]
```diff
@@ -55,7 +55,7 @@ private void Handle(MessagePayload<AdoptHero> obj)
                 return;
             }
 
-            ClanAdoptHeroPatch.RunFixedAdoptHero(adoptedHero, playerClan, playerHero);
+            //ClanAdoptHeroPatch.RunFixedAdoptHero(adoptedHero, playerClan, playerHero);
```
Comment (garrettluskey) [inline source/GameInterface/Services/Clans/Handlers/ClanAdoptionHandler.cs]: Can we delete this file?

File: source/GameInterface/Services/Clans/Handlers/ClanCompanionHandler.cs [hunk 4]
```diff
@@ -49,7 +49,7 @@ private void Handle(MessagePayload<AddCompanion> obj)
                 return;
             }
 
-            ClanAddCompanionPatch.RunOriginalAddCompanion(clan, companion);
+            //ClanAddCompanionPatch.RunOriginalAddCompanion(clan, companion);
```
Comment (garrettluskey) [inline source/GameInterface/Services/Clans/Handlers/ClanCompanionHandler.cs]: Can we delete this file?

File: source/GameInterface/Services/Clans/Handlers/ClanInfluenceHandler.cs [hunk 5]
```diff
@@ -43,7 +43,7 @@ private void Handle(MessagePayload<ChangeClanInfluence> obj)
                 return;
             }
 
-            ClanChangeInfluencePatch.RunOriginalChangeClanInfluence(clan, payload.Amount);
+            //ClanChangeInfluencePatch.RunOriginalChangeClanInfluence(clan, payload.Amount);
```
Comment (garrettluskey) [inline source/GameInterface/Services/Clans/Handlers/ClanInfluenceHandler.cs]: Can we delete this file?

File: source/GameInterface/Services/Clans/Handlers/ClanKingdomHandler.cs [hunk 6]
```diff
@@ -49,8 +49,8 @@ private void Handle(MessagePayload<ChangeClanKingdom> obj)
                 return;
             }
 
-            ClanChangeKingdomPatch.RunOriginalChangeClanKingdom(clan, newKingdom, (ChangeKingdomAction.ChangeKingdomActionDetail)payload.DetailId, 
-                payload.AwardMultiplier, payload.ByRebellion, payload.ShowNotification);
+            //ClanChangeKingdomPatch.RunOriginalChangeClanKingdom(clan, newKingdom, (ChangeKingdomAction.ChangeKingdomActionDetail)payload.DetailId,
```
Comment (garrettluskey) [inline source/GameInterface/Services/Clans/Handlers/ClanKingdomHandler.cs]: Can we delete this file?

File: source/GameInterface/Services/Clans/Handlers/ClanLeaderHandler.cs [hunk 7]
```diff
@@ -46,11 +46,11 @@ private void Handle(MessagePayload<ChangeClanLeader> obj)
             if (payload.NewLeaderId != null)
             {
                 objectManager.TryGetObject<Hero>(payload.NewLeaderId, out var newLeader);
-                ClanLeaderChangePatch.RunOriginalChangeClanLeader(clan, newLeader);
+                //ClanLeaderChangePatch.RunOriginalChangeClanLeader(clan, newLeader);
                 return;
             }
 
-            ClanLeaderChangePatch.RunOriginalChangeClanLeader(clan);
+            //ClanLeaderChangePatch.RunOriginalChangeClanLeader(clan);
```
Comment (garrettluskey) [inline source/GameInterface/Services/Clans/Handlers/ClanLeaderHandler.cs]: Can we delete this file?

File: source/GameInterface/Services/Clans/Handlers/ClanRenownHandler.cs [hunk 8]
```diff
@@ -42,7 +42,7 @@ private void Handle(MessagePayload<AddClanRenown> obj)
                 return;
             }
 
-            ClanAddRenownPatch.RunOriginalAddRenown(clan, payload.Amount, payload.ShouldNotify);
+            //ClanAddRenownPatch.RunOriginalAddRenown(clan, payload.Amount, payload.ShouldNotify);
```
Comment (garrettluskey) [inline source/GameInterface/Services/Clans/Handlers/ClanRenownHandler.cs]: Can we delete this file?

File: source/GameInterface/Services/Clans/Patches/ClanNameChangePatch.cs [hunk 9]
```diff
@@ -18,22 +14,29 @@ public class ClanNameChangePatch
     {
         private static ILogger Logger = LogManager.GetLogger<ClanNameChangePatch>();
 
-
         static bool Prefix(ref Clan __instance, TextObject name, TextObject informalName)
```
Comment (garrettluskey) [inline source/GameInterface/Services/Clans/Patches/ClanNameChangePatch.cs:17]: Can this file be removed since we are using autosync?

**Review by garrettluskey — CHANGES_REQUESTED** [review]

File: source/GameInterface/Services/Clans/Patches/ClanNameChangePatch.cs (hunk 9 above)
Comment (EgardA) [reply source/GameInterface/Services/Clans/Patches/ClanNameChangePatch.cs:17]: This cannot be removed as it's a call from client UI

Comment (EgardA) [reply source/GameInterface/Services/Clans/Patches/ClanNameChangePatch.cs:17]: Unless autosync has a way to permit client changes

**Review by garrettluskey — DISMISSED** [review]

**Comment by garrettluskey** [comment]
Please resolve conflicts then merge

**Pull request merged**